File: composio/cli/triggers.py
"""
Triggers manager for Composio SDK.

Usage:
    composio triggers [command] [options]
"""

import json
import typing as t

import click

from composio.cli.context import Context, pass_context
from composio.client.exceptions import NoItemsFound
from composio.exceptions import ComposioSDKError


@click.group(name="triggers", invoke_without_command=True)
@click.option(
    "--active",
    is_flag=True,
    default=False,
    help="Only list the active triggers",
)
@click.option(
    "--id",
    "trigger_ids",
    type=str,
    help="Filter by trigger id",
    multiple=True,
)
@click.option(
    "--app",
    "app_names",
    type=str,
    help="Filter by app name",
    multiple=True,
)
@pass_context
def _triggers(
    context: Context,
    trigger_ids: t.Tuple[str, ...],
    app_names: t.Tuple[str, ...],
    active: bool = True,
) -> None:
    """List triggers available for your account"""
    if context.click_ctx.invoked_subcommand:
        return

    try:
        if active or len(trigger_ids):
            triggers = context.client.active_triggers.get(trigger_ids=list(trigger_ids))
            if not triggers:
                raise click.ClickException("No active triggers found!")
            context.console.print("[green]Showing active triggers[/green]")
            for trg in triggers:
                context.console.print(f"• Name: {trg.triggerName}\n  ID: {trg.id}")
            return

        context.console.print("[green]Showing all triggers[/green]")
        for _trigger in context.client.triggers.get(app_names=list(app_names)):
            context.console.print(f"• {_trigger.name}")
    except ComposioSDKError as e:
        raise click.ClickException(message=e.message) from e


@_triggers.command(name="get")
@click.argument("id", type=str)
@pass_context
def _get(context: Context, id: str) -> None:
    """Get a specific trigger information."""
    context.console.print(
        f"\n[green]> Getting more details about trigger: {id}...[/green]\n"
    )
    try:
        (trigger,) = context.client.active_triggers.get(trigger_ids=[id])
        context.console.print(f"[bold]Trigger Name:[/bold] {trigger.triggerName}")
        context.console.print(f"[bold]Trigger ID:[/bold] {trigger.id}")
        context.console.print(f"[bold]Connection ID:[/bold] {trigger.connectionId}")
        context.console.print(
            f"[bold]Connection Config:[/bold] {json.dumps(trigger.triggerConfig, indent=2)}"
        )
        context.console.print(
            f"[bold]Disable this trigger using[/bold]: [red]composio-cli disable-trigger {id}[/red]"
        )
    except NoItemsFound as e:
        raise click.ClickException(
            message=(
                "No trigger found with the specified ID or it's not active; "
                "To list all active triggers, use the command: "
                "\n\tcomposio triggers --active"
            )
        ) from e
    except ComposioSDKError as e:
        raise click.ClickException(message=e.message) from e


@_triggers.command(name="enable")
@click.argument("id", type=str)
@pass_context
def _enable_trigger(context: Context, id: str) -> None:
    """Enable a trigger for an app"""
    context.console.print(f"Enabling trigger [green]{id}[/green]")
    try:
        (trigger,) = context.client.triggers.get(trigger_ids=[id])
        connected_account = context.client.get_entity().get_connection(
            app=trigger.appKey
        )

        config = {}
        properties = trigger.config.properties or {}
        for field in trigger.config.required or []:
            field_props = properties[field]
            field_title = field_props.title or field
            field_description = field_props.description or ""
            config[field] = click.prompt(text=f"{field_title} ({field_description})")

        response = context.client.triggers.enable(
            name=id,
            connected_account_id=connected_account.id,
            config=config,
        )
        context.console.print(
            f"Enabled trigger with ID: [green]{response['triggerId']}[/green]"
        )
    except ComposioSDKError as e:
        raise click.ClickException(message=e.message) from e


@_triggers.command(name="disable")
@click.argument("id", type=str)
@pass_context
def _disable_trigger(context: Context, id: str) -> None:
    """Disable a trigger for an app"""
    context.console.print(f"Disabling trigger [green]{id}[/green]")
    try:
        response = context.client.triggers.disable(id=id)
        if response["status"] == "success":
            context.console.print(f"Enabled trigger with ID: [green]{id}[/green]")
            return
        raise click.ClickException(f"Could not disable trigger with ID: {id}")
    except ComposioSDKError as e:
        raise click.ClickException(message=e.message) from e


@_triggers.group(name="callbacks")
def _callbacks() -> None:
    """Manage trigger callbacks."""


@_callbacks.command(name="set")
@click.argument("url", type=str)
@pass_context
def _set_callback(context: Context, url: str) -> None:
    """
    Set callback URL

    Note: Currently this command will set the provided URL as global callback URL
    """
    try:
        response = context.client.triggers.callbacks.set(url=url)
        context.console.print(response["message"])
    except ComposioSDKError as e:
        raise click.ClickException(message=e.message)


@_callbacks.command(name="get")
@pass_context
def _get_callback(context: Context) -> None:
    """
    Get callback URL
    """
    try:
        response = context.client.triggers.callbacks.get()
        context.console.print(response)
    except ComposioSDKError as e:
        raise click.ClickException(message=e.message)

File: js/src/cli/triggers.ts
/* eslint-disable no-console */
import chalk from "chalk";
import { Command } from "commander";
import client from "../sdk/client/client";

import inquirer from "inquirer";
import { Composio } from "../sdk";
import { getOpenAPIClient } from "../sdk/utils/config";

export default class ConnectionsCommand {
  private program: Command;

  constructor(program: Command) {
    this.program = program;

    const command = this.program.command("triggers");
    command
      .description("Manage and list triggers")
      .option("--id <text>", "Filter by trigger id")
      .option("--app <text>", "Filter by app name")
      .option("--active", "Show only active triggers")
      .action(async (options) => {
        if (options.active) {
          // Run active triggers command if --active flag is present
          const activeTriggers = new ActiveTriggers(command, false);
          // @ts-ignore
          await activeTriggers.handleAction();
        } else {
          // Otherwise run normal list action
          await this.handleAction(options);
        }
      });

    command
      .command("list")
      .description("List all triggers")
      .action(this.handleAction.bind(this));

    new TriggerAdd(command);
    new TriggerDisable(command);
    new ActiveTriggers(command);
    new TriggerEnable(command);
    new TriggerCallback(command);
  }

  private async handleAction(options: {
    active: boolean;
    id: string;
    app: string;
  }): Promise<void> {
    const { active, id, app } = options;
    const client = getOpenAPIClient();
    const { data, error } = await client.triggers.listTriggers({
      query: {
        ...(!!active && { showEnabledOnly: true }),
        ...(!!app && { appNames: app }),
        ...(!!id && { triggerIds: id }),
      },

      throwOnError: false,
    });

    if (error) {
      console.log(chalk.red((error as Error).message));
      return;
    }

    if (!data) {
      console.log(chalk.red("No triggers found"));
      return;
    }

    for (const trigger of data) {
      const typedTrigger = trigger;
      console.log(
        chalk.cyan(`  ${chalk.bold("Name")}:`),
        chalk.white(typedTrigger.appName)
      );
      console.log(
        chalk.cyan(`  ${chalk.bold("Enum")}:`),
        // @ts-ignore - typedTrigger.enum is not defined in the type but exists in the API response
        chalk.white(typedTrigger.enum)
      );
      console.log(
        chalk.cyan(`  ${chalk.bold("Description")}:`),
        chalk.white(typedTrigger.description)
      );
      console.log(""); // Add an empty line for better readability between triggers
    }
  }
}

export class TriggerAdd {
  private program: Command;

  constructor(program: Command) {
    this.program = program;

    this.program
      .command("add")
      .description("Add a new trigger")
      .argument("<trigger>", "The trigger name")
      .action(this.handleAction.bind(this));
  }

  async handleAction(triggerName: string): Promise<void> {
    const composioClient = new Composio({});

    const data = (await composioClient.triggers.list()).find(
      // @ts-ignore
      (trigger) => trigger.enum.toLowerCase() === triggerName.toLowerCase()
    );

    if (!data) {
      console.log(chalk.red(`Trigger ${triggerName} not found`));
      return;
    }

    const appName = data.appName;

    const connection = await composioClient
      .getEntity("default")
      .getConnection({ app: appName });

    if (!connection) {
      console.log(chalk.red(`Connection to app ${appName} not found`));
      console.log(
        `Connect to the app by running: ${chalk.cyan(`composio add ${appName}`)}`
      );
      return;
    }

    const dataConfig = data.config!;
    const properties = dataConfig.properties!;
    const requiredProperties = dataConfig.required! as string[];
    const configValue: Record<string, unknown> = {};

    for (const key in properties) {
      if (requiredProperties.includes(key)) {
        const prompt = inquirer.createPromptModule();
        const answer = await prompt([
          {
            type: "input",
            name: key,
            message: `Enter the value for ${key}`,
          },
        ]);

        configValue[key] = answer[key];
      }
    }

    const triggerSetupData = await composioClient.triggers.setup({
      connectedAccountId: connection.id,
      triggerName,
      config: configValue,
    });

    console.log(
      chalk.green(
        `Trigger ${triggerName} setup to app ${appName} with id ${triggerSetupData?.triggerId}`
      )
    );
  }
}

export class TriggerDisable {
  private program: Command;
  constructor(program: Command) {
    this.program = program;

    this.program
      .command("disable")
      .description("Disable an existing trigger")
      .argument("<triggerid>", "The trigger id")
      .action(this.handleAction.bind(this));
  }

  async handleAction(triggerId: string): Promise<void> {
    const composioClient = new Composio({});
    try {
      await composioClient.triggers.disable({ triggerId });
      console.log(chalk.green(`Trigger ${triggerId} disabled`));
    } catch (error) {
      console.log(chalk.red(`Error disabling trigger ${triggerId}: ${error}`));
    }
  }
}

export class TriggerEnable {
  private program: Command;
  constructor(program: Command) {
    this.program = program;

    this.program
      .command("enable")
      .description("Enable an existing trigger")
      .argument("<triggerid>", "The trigger id")
      .action(this.handleAction.bind(this));
  }

  async handleAction(triggerId: string): Promise<void> {
    const composioClient = new Composio({});
    try {
      await composioClient.triggers.enable({ triggerId });
      console.log(chalk.green(`Trigger ${triggerId} enabled`));
    } catch (error) {
      console.log(chalk.red(`Error enabling trigger ${triggerId}: ${error}`));
    }
  }
}

export class ActiveTriggers {
  private program: Command;
  constructor(program: Command, register: boolean = true) {
    this.program = program;

    if (register) {
      this.program
        .command("active")
        .description("Show list of currently active triggers")
        .action(this.handleAction.bind(this));
    }
  }

  async handleAction(): Promise<void> {
    const composioClient = new Composio({});
    const triggers = await composioClient.activeTriggers.list();
    for (const trigger of triggers) {
      console.log(`Id: ${chalk.bold(trigger.id)}`);
      console.log(`Trigger Name: ${chalk.cyan(trigger.triggerName)}`);
      console.log(
        `TriggerConfig: ${chalk.magenta(JSON.stringify(trigger.triggerConfig, null, 2))}`
      );
      console.log(`Connection ID: ${chalk.yellow(trigger.connectionId)}`);
      console.log(""); // Add an empty line for better readability between triggers
    }
  }
}

export class TriggerCallback {
  private program: Command;
  constructor(program: Command) {
    this.program = program;

    const callbackCommand = this.program
      .command("callback")
      .description("Manage trigger callback URLs");

    callbackCommand
      .command("set")
      .description("Set a callback URL for a trigger")
      .argument("<callbackURL>", "Callback URL that needs to be set")
      .action(this.handleSetAction.bind(this));

    callbackCommand
      .command("get")
      .description("Get the current callback URL for a trigger")
      .action(this.handleGetAction.bind(this));
  }

  async handleSetAction(callbackURL: string): Promise<void> {
    getOpenAPIClient();
    try {
      await client.triggers.setCallbackUrl({
        body: {
          callbackURL: callbackURL,
        },
      });
      console.log(chalk.green(`Callback URL set to ${callbackURL}`));
    } catch (error) {
      console.log(
        chalk.red(
          `Error setting callback URL to ${callbackURL}: ${(error as Error).message}`
        )
      );
    }
  }

  async handleGetAction(): Promise<void> {
    getOpenAPIClient();
    try {
      const res = await client.triggers.getWebhookUrl();
      console.log(
        chalk.green(`Current callback URL is ${res?.data?.callbackURL}`)
      );
    } catch (error) {
      console.log(
        chalk.red(`Error getting callback URL: ${(error as Error).message}`)
      );
    }
  }
}
